WIP: Add 3D (STL) file support

tschwemley opened the pull request:
This ticket is currently a work in progress.
- [ ] Update file validation on client side
- [ ] Update file validation on server side
- [ ] Get STL preview showing in pre-publish view
- [ ] Validate that upload works and isn't showing a false-positive message

This adds support for STL file uploads (and preview) to the Spee.ch platform.

skhameneh commented on client/src/utils/file.js:
```diff
@@ -26,7 +29,23 @@ module.exports = {
+        // Handle STL check (file.type is empty for STLs and file.type cannot be reassigned)
+        file.isStl = false;
+        return new Promise((resolve, reject) => {
+          const loader = new FileReader();
```
This entangles synchronous validation with a threejs loader that uses async events.
Being a synchronous method, this shouldn't wrap with a promise either.

skhameneh commented on client/src/utils/file.js:
```diff
@@ -26,7 +29,23 @@ module.exports = {
+          loader.readAsArrayBuffer(file);
+        })
+          .catch(() => {
+            throw new Error(file.type + ' is not a supported file type. Only, .jpeg, .png, .gif, and .mp4 files are currently supported.');
```
This error needs to be updated with this PR (add STL).

skhameneh commented on client/src/utils/stl.js:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Module to determine if file upload is a valid stl file. Not using ES6 class since server isn't set up to compile
```
This comment (about ES6 classes) can be removed and ES6 classes are supported in Node.
The majority of ES6 features are supported by Node with the exception of modules being under experimental support.

tschwemley replied:
In this case it wasn't compiling but I see your point on the comment since isn't explicitly correct. The intent was to explain why I opted to go for an object export rather than using a class. In hindsight the comment was unnecessary and could have been noted in a comment.

skhameneh commented on client/src/utils/file.js:
```diff
@@ -26,7 +29,23 @@ module.exports = {
+          };
+          loader.readAsArrayBuffer(file);
+        })
+          .catch(() => {
```
Adjust indentation

skhameneh commented on server/controllers/api/claim/publish/validateFileTypeAndSize.js:
```diff
@@ -24,8 +25,24 @@ const validateFileTypeAndSize = (file) => {
+      })
+        .catch(() => {
+          logger.debug('publish > file validation > unrecognized file type');
+          throw new Error('The ' + file.type + ' content type is not supported.  Only, image/jpg, image/png, image/gif, and video/mp4 content types are currently supported.');
```
Update error to include STL

skhameneh commented on server/controllers/api/claim/publish/validateFileTypeAndSize.js:
```diff
@@ -24,8 +25,24 @@ const validateFileTypeAndSize = (file) => {
-      throw new Error('The ' + file.type + ' content type is not supported.  Only, image/jpg, image/png, image/gif, and video/mp4 content types are currently supported.');
+      // Handle STL check (file.type is empty for STLs and file.type cannot be reassigned)
+      file.isStl = false;
+      return new Promise((resolve, reject) => {
```
This entangles synchronous validation with a threejs loader that uses async events.
Being a synchronous method, this shouldn't wrap with a promise either.

skhameneh commented on client/src/components/ThreeScene/index.jsx:
```diff
@@ -0,0 +1,74 @@
+    reader.onloadend = (e) => {
+      const loader = new STLLoader();
+      loader.load(e.target.result, (geometry) => {
+        console.log(geometry);
```
Can remove this 😃

skhameneh commented on client/src/components/ThreeScene/index.jsx:
```diff
@@ -0,0 +1,74 @@
+        mesh.castShadow = true;
+        mesh.receiveShadow = true;
+        this.scene.add(mesh);
+        console.log('in here');
```
Remove 😃

skhameneh commented on client/src/components/ThreeScene/index.jsx:
```diff
@@ -0,0 +1,74 @@
+    this.renderer.setPixelRatio(window.devicePixelRatio);
+    this.renderer.setSize(width, height);
+
+    this.mount.appendChild(this.renderer.domElement);
```
You might want to check the PR's from @btzr-io here https://github.com/lbryio/lbry-desktop/pull/1558 and get his feedback as well; he's created some reusable components for 3D viewing that might give some inspiration.

tschwemley commented:
I have gone ahead and cleaned up the validation by making everything async instead of mixing async and sync validation. I have also added additional limitations to the STL file uploads (10mb limit and 1,000,000 poly limit). I chose the file size limit arbitrarily and based the poly limit off of augment.com guidelines (http://www.augment.com/help/requirements-and-limitations/). I can alter these values as needed if there is any feedback.

Currently wrapping up on the preview, then hopefully we'll be close to wrapping this up!

skhameneh commented:
Hey @tschwemley, you left of with a mention of wrapping things up, where are you with this?

tzarebczan commented:
@tschwemley, are you still interested in finishing this up?
